sdk/servicebus/azure-messaging-servicebus/src/test/java/com/azure/messaging/servicebus/ServiceBusReceiverAsyncClientIntegrationTest.java

```diff
@@ -511,40 +531,112 @@ void peekMessages(MessagingEntityType entityType, boolean isSessionEnabled) {
+                        && !receivedPositions.parallelStream().filter(position ->
+                            position.intValue() == (int) receivedMessage.getApplicationProperties().get(MESSAGE_POSITION_ID))
+                        .findFirst().isPresent()
+                        && (int) receivedMessage.getApplicationProperties().get(MESSAGE_POSITION_ID) >= 3
+                        && (int) receivedMessage.getApplicationProperties().get(MESSAGE_POSITION_ID) <= 6)
+                    .flatMap(receivedMessage -> {
+                        receivedPositions.add((Integer) receivedMessage.getApplicationProperties().get(MESSAGE_POSITION_ID));
+                        actualCount.incrementAndGet();
+                        return Mono.just(receivedMessage);
+                    })
+                    .repeat(() -> actualCount.get() < 4)
+                    .collectSortedList(new Comparator<ServiceBusReceivedMessage>() {
+                        @Override
+                        public int compare(ServiceBusReceivedMessage o1, ServiceBusReceivedMessage o2) {
+                            int position1 = (int) o1.getApplicationProperties().get(MESSAGE_POSITION_ID);
+                            int position2 = (int) o2.getApplicationProperties().get(MESSAGE_POSITION_ID);
+                            return position1 - position2;
+                        }
+                    })
+                    .block();
+                assertEquals(4, actualCount.get(), "Failed to peek four messages");
+                receivedMessages.forEach(actualMessage -> checkCorrectMessage.accept(actualMessage, messageCount.getAndIncrement()));
+            }));
+
+            threadList.add(new Thread(() -> {
+                AtomicLong actualCount = new AtomicLong();
+                List<ServiceBusReceivedMessage> receivedMessages
+                    = receiver.peekMessage(sessionId)
+                    .filter(receivedMessage -> messageId.equals(receivedMessage.getMessageId())
+                        && (int) receivedMessage.getApplicationProperties().get(MESSAGE_POSITION_ID) == 7)
+                    .flatMap(receivedMessage -> {
+                        actualCount.incrementAndGet();
+                        return Mono.just(receivedMessage);
+                    })
+                    .repeat(() -> actualCount.get() < 1)
+                    .collectList()
+                    .block();
+                assertEquals(1, actualCount.get(), "Failed to peek message only one");
+                receivedMessages.forEach(actualMessage -> checkCorrectMessage.accept(actualMessage, 7));
+            }));
+
+            threadList.parallelStream().forEach(t -> {
+                t.start();
+                try {
+                    t.join(OPERATION_TIMEOUT.toMillis());
+                } catch (InterruptedException e) {
+                    e.printStackTrace();
+                }
```

Comment:
Which thread does this join happen?

I assume `join` should happen at the thread of the test case. But I am not sure which thread runs this in `parallelStream`.

And what happens if `join` pass the timeout?

Reply:
Or should you use `ExecutorService` instead list of threads?

Reply:
1. When we invoke the join() method on test thread, the test thread goes into a waiting state. It remains in a waiting state until the referenced thread terminates. If the referenced thread is blocked or is taking too long to process. `The operation was canceled` exception will occur on the test thread. 
2.  join() with Timeout means to waits at most millis milliseconds for this thread to die.